pkg/controllermanager/controller/managedseedset/managedseedset_actuator.go

```diff
@@ -0,0 +1,666 @@
+}
+
+func shootReconcileFailed(shoot *gardencorev1beta1.Shoot) bool {
+	lastOp := shoot.Status.LastOperation
+	return shoot.Generation == shoot.Status.ObservedGeneration && lastOp != nil &&
+		(lastOp.Type == gardencorev1beta1.LastOperationTypeCreate || lastOp.Type == gardencorev1beta1.LastOperationTypeReconcile) &&
+		lastOp.State == gardencorev1beta1.LastOperationStateFailed
+}
+
+func shootDeleteFailed(shoot *gardencorev1beta1.Shoot) bool {
+	lastOp := shoot.Status.LastOperation
+	return shoot.Generation == shoot.Status.ObservedGeneration && lastOp != nil &&
+		lastOp.Type == gardencorev1beta1.LastOperationTypeDelete &&
+		lastOp.State == gardencorev1beta1.LastOperationStateFailed
+}
+
+func managedSeedRegistered(managedSeed *seedmanagementv1alpha1.ManagedSeed) bool {
+	conditionSeedRegistered := gardencorev1beta1helper.GetCondition(managedSeed.Status.Conditions, seedmanagementv1alpha1.ManagedSeedSeedRegistered)
+	return managedSeed.Generation == managedSeed.Status.ObservedGeneration && managedSeed.DeletionTimestamp == nil &&
+		conditionSeedRegistered != nil && conditionSeedRegistered.Status == gardencorev1beta1.ConditionTrue
+}
+
+func seedReady(seed *gardencorev1beta1.Seed) bool {
+	conditionGardenletReady := gardencorev1beta1helper.GetCondition(seed.Status.Conditions, gardencorev1beta1.SeedGardenletReady)
+	conditionBootstrapped := gardencorev1beta1helper.GetCondition(seed.Status.Conditions, gardencorev1beta1.SeedBootstrapped)
+	conditionBackupBucketsReady := gardencorev1beta1helper.GetCondition(seed.Status.Conditions, gardencorev1beta1.SeedBackupBucketsReady)
+	return seed.Generation == seed.Status.ObservedGeneration && seed.DeletionTimestamp == nil &&
+		conditionGardenletReady != nil && conditionGardenletReady.Status == gardencorev1beta1.ConditionTrue &&
+		conditionBootstrapped != nil && conditionBootstrapped.Status == gardencorev1beta1.ConditionTrue &&
+		(conditionBackupBucketsReady == nil || conditionBackupBucketsReady.Status == gardencorev1beta1.ConditionTrue)
+}
+
+func shootHealthStatus(shoot *gardencorev1beta1.Shoot) operationshoot.Status {
+	if value, ok := shoot.Labels[common.ShootStatus]; ok {
+		return operationshoot.Status(value)
+	}
+	return operationshoot.StatusProgressing
+}
+
+// getName returns the replica object name for the given set and ordinal.
+func getName(set *seedmanagementv1alpha1.ManagedSeedSet, ordinal int) string {
+	return fmt.Sprintf("%s-%d", set.Name, ordinal)
+}
+
+// parentNameAndOrdinalRegex is a regular expression that extracts the parent name and ordinal from the name of a replica object
+var parentNameAndOrdinalRegex = regexp.MustCompile("(.*)-([0-9]+)$")
+
+// getParentNameAndOrdinal gets the parent name and ordinal from the given replica object name.
+// If the object was not created by a ManagedSeedSet, its parent name is considered to be empty string,
+// and its ordinal is considered to be -1.
+func getParentNameAndOrdinal(name string) (string, int) {
```

Comment:
I don't see any usage of the first return parameter, yet.
For what do you think, you will need it?
In general, the "parent name" should be determined by looking at the `ownerReferences`.

Reply:
No, I don't think this is needed, so I removed it.